### ISSUE TITLE
[Feature Request] Add MemberSerialization option to JsonSerializerSettings

### ISSUE THREAD

**Issue by doobah** [JamesNK/Newtonsoft.Json#2506]
I'd like to be able to customize member serialization to be opt in by default. That is, instead of annotating every class with `[JsonObject(MemberSerialization = MemberSerialization.OptIn)]`
I'd like to just set it once in the serializer, e.g.

```
JsonConvert.DefaultSettings = delegate
{
    JsonSerializerSettings settings = new JsonSerializerSettings();
    settings.MemberSerialization = MemberSerialization.OptIn;

    return settings;
};
```

This would make its usage similar to the MissingMemberHandling option which exists in both JsonSerializerSettings and JsonObjectAttribute.

Is there any reason why this couldn't/shouldn't be done? I'm happy to make a pull request.

**Comment by elgonzo** [comment]
I have difficulties understanding the motivation/purpose behind this. Help me understand.
Unless i am mistaken, for OptIn to make any sense it requires the participating properties/fields to be decorated with `[JsonProperty]`/`[DataMember]` attributes, no? Thus, since you have to annotate all classes participating in (de)serialization with a dose of `[JsonProperty]`/`[DataMember]` attributes anyway, why the very specific dislike of annotating the classes with [JsonObject] as well?

**Comment by elgonzo** [comment]
> To be clear, I'm completely fine with annotating with [JsonObject], I'd just rather not have to set MemberSerialization = MemberSerialization.OptIn every time.

Maybe you could go for `[DataContract]`/`[DataMember]` instead of `[JsonProperty]` attributes then (unless you have to rely on `[JsonProperty]`-specific parameters, of course). Json.NET recognizes the `[DataContract]`/`[DataMember]` attributes, and using these two attributes will automatically employ opt-in behavior (i.e., in a `[DataContract]`-annotated class, only properties annotated with `[DataMember]` will participate in (de)serialization).

**Comment by doobah** [comment]
Yeah I think some of the ``JsonProperty`` parameters could come in handy, for example I might want to set ``ItemTypeNameHandling`` for a list that contains polymorphic objects. For this reason I'd rather not use ``DataContract``.

I don't think ``JsonIgnore`` will help me much either since in my use case it's quite common that most of the members should not be serialised. I'm developing a system that includes classes that provide an interface to binary files, e.g. bitmaps or audio clips. The classes contain a bunch of members that are populated with the binary data or metadata once the file has been 'loaded', but the only bit of data that needs to be serialized  is the path to the binary file.

Is my use-case uncommon?

**Comment by elgonzo** [comment]
If you have to stick with `[JsonProperty]`, there is still another possible way for you avoiding to specify `MemberSerialization = MemberSerialization.OptIn` by using a custom contract resolver.

Derive such a custom contract resolver from `DefaultContractResolver`, and all you would (probably) need to do is to override its `GetSerializableMembers` method to select only properties from a given type that are annotated with the `[JsonProperty]` attribute. It could be as basic as:
```
public class OptInContractResolver : DefaultContractResolver
{
    protected override List<MemberInfo> GetSerializableMembers(Type objectType)
    {
        return new List<MemberInfo>(
            objectType.GetProperties(BindingFlags.Instance | BindingFlags.Public)
                .Where(m => m.IsDefined(typeof(JsonPropertyAttribute)))
        );
    }
}
```

(Depending on your specific circumstances, you might constrain this custom behavior of _GetSerializableMembers_ to your own data types, for example by calling _base.GetSerializableMembers_ for all types that are not from your program assembly/ies.)

Code example demonstrating this: https://dotnetfiddle.net/ixvkYq

**Comment by elgonzo** [comment]
If you have to stick with `[JsonProperty]`, there is still another possible way for you avoiding to specify `MemberSerialization = MemberSerialization.OptIn` in `[JsonObject]` attributes by using a custom contract resolver that realizes the opt-in behavior.

Derive such a custom contract resolver from `DefaultContractResolver`, and all you would (probably) need to do is to override its `GetSerializableMembers` method to select only properties from a given type that are annotated with the `[JsonProperty]` attribute. It could be as basic as:
```
public class OptInContractResolver : DefaultContractResolver
{
    protected override List<MemberInfo> GetSerializableMembers(Type objectType)
    {
        return new List<MemberInfo>(
            objectType.GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.FlattenHierarchy)
                .Where(m => m.IsDefined(typeof(JsonPropertyAttribute)))
        );
    }
}
```

(Depending on your specific circumstances, you might constrain this custom behavior of _GetSerializableMembers_ to your own data types.)

Code example demonstrating this: https://dotnetfiddle.net/ixvkYq

**Comment by doobah** [comment]
Oh that looks pretty useful! I'll give this a go

**Comment by elgonzo** [comment]
I just noticed a flaw in my example on dotnetfiddle. `DefaultContractResolver` normally caches the contracts of resolved types, thus the processing of the contract resolver should only be a one-time cost per type involved in the (de)serialization.

However, the `JsonConvert.DefaultSettings` delegate is executed again for every single `JsonConvert.SerializeObject`, `JsonConvert.DeserializeObject` and `JsonSerializer.CreateDefault` invocation. Since in my example the contract resolver is  instantiated as part of the delegate code, my example basically renderes the caching done by the contract resolver ineffectual. Solution to this problem: Create the contract resolver instance - or the whole JsonSerializerSettings object - only once outside the `JsonConvert.DeserializeObject` delegate code...

My apologies for this oversight...

**Comment by pointcache** [comment]
Add it please. This should have always been there.
I dont want to think if i forgot or not to add OptIn to some obscure class, so that the whole game breaks because "i dont get why you need it". How can it even be that such things are left to assumptions that nobody will need it? 
Ok lets say its not gamedev, you have a product with a bunch of third party compiled libraries and types, how do you handle this case without global optin? Everything in them will be serialized, every property will be called, everything will be dumped in the json. Just what.

**Comment by MGRussell** [comment]
I'd also like to see this for the sake of a global opt out. When inheriting from classes that have the [DataContract] attribute, it seems that attribute is considered inherited by Json.net which leads to serialization behavior that can vary depending on the attributes of parent classes (since the presence of the [DataContract] attribute results in opt-in serialization, against the default of opt-out), which may be subject to change. 

Of course this can also be achieved through a contract resolver but it's somewhat more obtuse.

**Comment by elgonzo** [comment]
> Json.NET's default behavior swaps from opt-out to opt-in if certain attributes like [DataContract] are present.

It has to use opt-in for [DataContract]/[DataMember], because that is the [specified and documented behavior for these attributes](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractattribute#remarks). Not doing so would go against the documented behavior of these attributes.

> [...] so attributes anywhere in the inheritance tree (which of course can change) can change the serialization behavior of a class. I don't find this behavior desirable.

If you don't like the by-definition behavior of [DataContract]/[DataMember] being opt-in, then don't use these attributes. Use [JsonObject]/[JsonProperty] attributes instead. The only reason to avoid attributes specific to Newtonsoft.Json is wanting to (potentially) use other serializers instead of Newtonsoft.Json. And guess what: Expect that other serializers will also follow the opt-in behavior as demanded byt [DataContract]/[DataMember] -- though luck finding any serializer that also offers a [DataContract]/[DataMember] behavior that goes against the documented behavior.

**Comment by MGRussell** [comment]
I'd recommend reading your own links before trying to correct others, especially when acting like a tool while doing so.

> A data contract differs significantly from an actual interface in its inheritance behavior. Interfaces are inherited by any derived types. When you apply the [DataContractAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractattribute?view=net-6.0) to a base class, the derived types do not inherit the attribute or the behavior.

This is an unintended (and known) bug in Json.NET that was maintained for the sake of backwards compatibility.

**Comment by elgonzo** [comment]
> especially when acting like a tool while doing so

Acting like a tool? My, my, making friends must come quite natural to you... :grin:

**Comment by Edgaras91** [comment]
I want to set `MemberSerialization.Fields` without adding a single attribute on any models/classes. I tried writing my serializer with some success, but the code is far from how the Newtonsoft library does it.
Any suggestions on how I can implement this, or if this feature is in the works?

**Comment by Edgaras91** [comment]
Also my use case:
I have a huge old monolith web application, that has lots of models, and some being returned from controller to front end, using Newtonsoft serialization, for which the models don't have any Json attributes on them yet.

I have now a requirement to introduce distributed caching (Redis) and I need to serialize the data using Json serializer.
The same models used in controllers are also being cached.
If I will add the needed `[JsonObject(MemberSerialization.Fields)]` on lots of models, it means we will be returning private field data in our web response models, when the scope is to serialize the fields for these models only for caching.

Adding a global setting (`JsonSerializerSettings`) would allow me to serialize just for distributed caching, and not affect existing model serialization elsewhere (controllers)
